runtime/lua/vim/filetype/detect.lua

```diff
@@ -315,7 +783,44 @@ function M.redif(bufnr)
   end
 end
 
-function M.rules(path, bufnr) end
+local udev_rules_pattern = '^%s*udev_rules%s*=%s*"([%^"]+)/*".*'
+function M.rules(path, bufnr)
+  -- TODO: better variable name?
+  local path_ = vim.fn.expand(path, ':p')
+  -- The third pattern also matches "/usr/lib/udev/*.rules" so Vim's implementation was simplified here
+  if
+    findany(path_, {
+      '/etc/udev/.*%.rules$',
+      '/etc/udev/rules%.d/.*$.rules$',
+      '/lib/udev/.*%.rules$',
+      '/lib/udev/rules%.d/.*%.rules$',
+    })
+  then
+    vim.bo[bufnr].filetype = 'udevrules'
+  elseif path_:find('^/etc/ufw/') then
+    -- Better than hog
+    vim.bo[bufnr].filetype = 'conf'
+  elseif findany(path_, { '^/etc/polkit%-1/rules%.d', '/usr/share/polkit%-1/rules%.d' }) then
+    vim.bo[bufnr].filetype = 'javascript'
+  else
+    local ok, config_lines = pcall(vim.fn.readfile, '/etc/udev/udev.conf')
+    if not ok then
+      vim.bo[bufnr].filetype = 'hog'
+      return
+    end
+    local dir = vim.fn.expand(path, ':p:h')
+    for _, line in ipairs(config_lines) do
+      local match = line:match(udev_rules_pattern)
+      local udev_rules = line:gsub(udev_rules_pattern, match, 1)
+      if dir == udev_rules then
+        vim.bo[bufnr].filetype = 'udevrules'
+        -- TODO: pretty sure Vim's break is incorrect here?
```

Comment:
Probably, but it still works since they use `setf` so a subsequent `setf` will have no effect. Returning here is the right call though.